test/src/compose/images.spec.ts

```diff
@@ -0,0 +1,481 @@
+				await imageManager.remove(imageToRemove);
+
+				// This checks that the remove method was ultimately called
+				expect(mockerode.removeImage).to.have.been.calledOnceWith(
+					imageToRemove.name,
+				);
+
+				// Check that the image was removed from the db
+				expect(await testDb.models('image').select().where(imageToRemove)).to.be
+					.empty;
+			},
+			{ images },
+		);
+	});
+
+	it('removes image from DB and engine when there is a single DB image with matching dockerImageId', async () => {
+		// Newer image
+		const imageToRemove = {
+			id: 246,
+			name:
+				'registry2.balena-cloud.com/v2/793f9296017bbfe026334820ab56bb3a@sha256:2c969a1ba1c6bc10df53481f48c6a74dbd562cfb41ba58f81beabd03facf5582',
+			appId: 1658654,
+			serviceId: 650325,
+			serviceName: 'app_1',
+			imageId: 2693229,
+			releaseId: 1524186,
+			dependent: 0,
+			dockerImageId:
+				'sha256:f1154d76c731f04711e5856b6e6858730e3023d9113124900ac65c2ccc90e8e7',
+		};
+
+		// Insert images into the db
+		await testDb.models('image').insert([
+			imageToRemove,
+			{
+				id: 247,
+				name:
+					'registry2.balena-cloud.com/v2/902cf44eb0ed51675a0bf95a7bbf0c91@sha256:12345a1ba1c6bc10df53481f48c6a74dbd562cfb41ba58f81beabd03facf5582',
+				appId: 1658654,
+				serviceId: 650331,
+				serviceName: 'app_2',
+				imageId: 2693230,
+				releaseId: 1524186,
+				dependent: 0,
+				dockerImageId:
+					'sha256:f1154d76c731f04711e5856b6e6858730e3023d9113124900ac65c2ccc901234',
+			},
+		]);
+
+		// Engine image state
+		const images = [
+			// The image to remove
+			createImage(
+				{
+					Id:
+						'sha256:f1154d76c731f04711e5856b6e6858730e3023d9113124900ac65c2ccc90e8e7',
+				},
+				{
+					// The target digest matches the image name
+					References: [
+						'registry2.balena-cloud.com/v2/793f9296017bbfe026334820ab56bb3a:delta-ada9fbb57d90e61e@sha256:2c969a1ba1c6bc10df53481f48c6a74dbd562cfb41ba58f81beabd03facf5582',
+					],
+				},
+			),
+			// Other images to test
+			createImage(
+				{
+					Id: 'aaa',
+				},
+				{
+					References: ['balena/aarch64-supervisor:11.11.11'],
+				},
+			),
+			// The other image on the database
+			createImage(
+				{
+					Id:
+						'sha256:f1154d76c731f04711e5856b6e6858730e3023d9113124900ac65c2ccc901234',
+				},
+				{
+					References: [
+						'registry2.balena-cloud.com/v2/902cf44eb0ed51675a0bf95a7bbf0c91:delta-80ed841a1d3fefa9@sha256:12345a1ba1c6bc10df53481f48c6a74dbd562cfb41ba58f81beabd03facf5582',
+					],
+				},
+			),
+		];
+
+		// Perform the test with our specially crafted data
+		await withMockerode(
+			async (mockerode) => {
+				// Check that the image exists
+				// this is really checking that mockerode works ok
+				await expect(
+					mockerode.getImage(imageToRemove.name).inspect(),
+					'image exists on the engine before test',
+				).to.not.be.rejected;
+
+				// Check that only one image with this dockerImageId exists in the db
+				// in memory db is a bit flaky sometimes, this checks for issues
```

Comment:
> in memory db is a bit flaky

How so?

Reply:
I'm not sure, that module failed one of the tests randomly because of the database entry not being there, which doesn't makes sense because the entry is being inserted a couple of lines before.